components/style/matching.rs

```diff
@@ -769,22 +769,27 @@ pub trait MatchMethods : TElement {
                 //
                 //  2. This is an incremental restyle, but the old display value
                 //     is none, so there's no effective way for Gecko to get the
-                //     style source. In this case, we could return either
+                //     style source (which is the style context).
+                //
+                //     In this case, we could return either
                 //     RestyleDamage::empty(), in the case both displays are
                 //     none, or rebuild_and_reflow, otherwise. The first case
                 //     should be already handled when calling this function, so
                 //     we can assert that the new display value is not none.
                 //
-                //     Also, this can be a text node (in which case we don't
-                //     care of watching the new display value).
+                //     Also, note that some styled pseudo-elements may not have
+                //     a frame even when the display value is other than none,
+                //     so we need to actually compare both values here.
```

Comment:
Can you give an example of what pseudos, this happens for, and why? Does this basically mean that we need to generate RECONSTRUCT_FRAME for every such pseudo-element when its style changes? Is that what Gecko does?